apps/aecore/src/aec_chain_state.erl

```diff
@@ -279,9 +285,25 @@ internal_insert_1(Node, Original, State) ->
 check_update_after_insert(Node, State) ->
     case determine_chain_relation(Node, State) of
         off_chain -> State;
-        new_top -> update_state_tree(Node, maybe_add_genesis_hash(State, Node));
-        in_chain -> update_state_tree(Node, State);
-        {fork, ForkNode} -> update_state_tree(ForkNode, State)
+        new_top ->
+            State1 = update_state_tree(Node, maybe_add_genesis_hash(State, Node)),
+            case find_top_header_hash(State1) of
+                not_found -> State1;
+                {ok, TopHeaderHash, TopDifficulty} ->
+                    State2 = set_top_header_hash(TopHeaderHash, State1),
+                    set_top_header_difficulty(TopDifficulty, State2)
+            end;
+        in_chain ->
+            %% This cannot change the top header
+            update_state_tree(Node, State);
+        {fork, ForkNode} ->
+            State1 = update_state_tree(ForkNode, State),
+            case find_top_header_hash(State1) of
+                not_found -> State1;
+                {ok, TopHeaderHash, TopDifficulty} ->
+                    State2 = set_top_header_hash(TopHeaderHash, State1),
+                    set_top_header_difficulty(TopDifficulty, State2)
```

Comment:
I see these (hash and diff.) are set separately in map though persisted atomically. Ok.